torchmultimodal/modules/encoders/mdetr_image_encoder.py

```diff
@@ -0,0 +1,146 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+from typing import Any, Dict, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import nn, Tensor
+from torchvision.models._utils import IntermediateLayerGetter
+from torchvision.models.resnet import resnet101, ResNet101_Weights
+
+
+class FrozenBatchNorm2d(nn.Module):
+    """
+    BatchNorm2d where the batch statistics and the affine parameters are fixed.
+
+    Copy-paste from torchvision.misc.ops with added eps before rqsrt,
+    without which any other models than torchvision.models.resnet[18,34,50,101]
+    produce nans.
+    """
+
+    def __init__(self, n):
+        super().__init__()
+        self.register_buffer("weight", torch.ones(n))
+        self.register_buffer("bias", torch.zeros(n))
+        self.register_buffer("running_mean", torch.zeros(n))
+        self.register_buffer("running_var", torch.ones(n))
+
+    def _load_from_state_dict(
+        self,
+        state_dict,
+        prefix,
+        local_metadata,
+        strict,
+        missing_keys,
+        unexpected_keys,
+        error_msgs,
+    ):
+        num_batches_tracked_key = prefix + "num_batches_tracked"
+        if num_batches_tracked_key in state_dict:
+            del state_dict[num_batches_tracked_key]
+
+        super()._load_from_state_dict(
+            state_dict,
+            prefix,
+            local_metadata,
+            strict,
+            missing_keys,
+            unexpected_keys,
+            error_msgs,
+        )
+
+    def forward(self, x: torch.Tensor):
+        # move reshapes to the beginning to make it fuser-friendly
+        # ignore mypy errors because fixing them would break checkpoint loading
+        w = self.weight.reshape(1, -1, 1, 1)  # type: ignore
+        b = self.bias.reshape(1, -1, 1, 1)  # type: ignore
+        rv = self.running_var.reshape(1, -1, 1, 1)  # type: ignore
+        rm = self.running_mean.reshape(1, -1, 1, 1)  # type: ignore
+        eps = 1e-5
+        scale = w * (rv + eps).rsqrt()
+        bias = b - rm * scale
+        return x * scale + bias
+
+
+class PositionEmbeddingSine(nn.Module):
```

Comment:
neither used nor tested?

Reply:
Not tested explicitly here, but its used in #77 (and tested there by loading and calling forward on the full model). In the original repo they tie this to the backbone, but I decided to keep it separate. But it is still in here as a holdover. Happy to move it elsewhere if you think that's clearer

Reply:
maybe call it 2dposition embeddings